### PR TITLE
Cherrypick fixes for 23.0-rc2 into 23.x

### PR DESCRIPTION
**Pull request by deannagarcia**

### CODE REVIEW

File: src/google/protobuf/descriptor.cc [hunk 1]
```diff
@@ -2612,10 +2612,12 @@ void FileDescriptor::CopyHeadingTo(FileDescriptorProto* proto) const {
   }
```
Comment (mkruskal-google) [inline src/google/protobuf/descriptor.cc:2613]: You're missing https://github.com/protocolbuffers/protobuf/commit/9b218c1d9aaa8108f65cca0d65325936a8bbeac6 I think

**Comment by mkruskal-google** [comment]
Should we just pull in every change on main to be safe?  It's only a day, and I'm worried we might be missing dependent commits

File: src/google/protobuf/descriptor.cc (hunk 1 above)
Comment (deannagarcia) [reply src/google/protobuf/descriptor.cc:2615]: Pulled it in, thanks for the catch!

**Comment by deannagarcia** [comment]
> Should we just pull in every change on main to be safe? It's only a day, and I'm worried we might be missing dependent commits

Discussed offline, tools work better with these cherrypicks so we are going to stick with this strategy.

**Review by fowles — APPROVED** [review]

**Review by mkruskal-google — APPROVED** [review]

**Pull request merged**